npe2/_fetch.py

```diff
@@ -0,0 +1,269 @@
+from __future__ import annotations
+
+import json
+import site
+import sys
+import tempfile
+import warnings
+from contextlib import contextmanager
+from functools import lru_cache
+from importlib import metadata
+from io import BytesIO
+from logging import getLogger
+from pathlib import Path
+from typing import TYPE_CHECKING, Any, Dict, Iterator, List, Optional
+from urllib.request import urlopen
+from zipfile import ZipFile
+
+from build.env import IsolatedEnvBuilder
+
+if TYPE_CHECKING:
+    import build.env
+
+    from npe2.manifest.schema import PluginManifest
+
+logger = getLogger(__name__)
+
+NPE1_ENTRY_POINT = "napari.plugin"
+NPE2_ENTRY_POINT = "napari.manifest"
+__all__ = [
+    "fetch_manifest",
+    "get_pypi_url",
+    "isolated_plugin_env",
+    "get_hub_plugin",
+    "get_hub_plugins",
+]
+
+
+def fetch_manifest(package: str, version: Optional[str] = None) -> PluginManifest:
+    """Fetch a manifest for a pip specifier (package name, possibly with version).
+
+    Parameters
+    ----------
+    package : str
+        package name
+    version : Optional[str]
+        package version, by default, latest version.
+
+    Returns
+    -------
+    PluginManifest
+        Plugin manifest for package `specifier`.
+    """
+    from npe2 import PluginManifest
+    from npe2.manifest import PackageMetadata
+
+    has_npe1 = False
+
+    # first just grab the wheel from pypi with no dependencies
+    try:
+        with _tmp_pypi_wheel_download(package, version) as td:
+            # create a PathDistribution from the dist-info directory in the wheel
+            dist = metadata.PathDistribution(next(Path(td).glob("*.dist-info")))
+
+            for ep in dist.entry_points:
+                # if we find an npe2 entry point, we can just use
+                # PathDistribution.locate_file to get the file.
+                if ep.group == NPE2_ENTRY_POINT:
+                    logger.debug("pypi wheel has npe2 entry point.")
+                    # python 3.8 fallbacks
+                    match = ep.pattern.match(ep.value)
+                    assert match
+                    module: str = match.groupdict()["module"]
+                    attr: str = match.groupdict()["attr"]
+
+                    mf_file = dist.locate_file(Path(module) / attr)
+                    mf = PluginManifest.from_file(str(mf_file))
+                    # manually add the package metadata from our distribution object.
+                    mf.package_metadata = PackageMetadata.from_dist_metadata(
+                        dist.metadata
+                    )
+                    return mf
+                elif ep.group == NPE1_ENTRY_POINT:
+                    has_npe1 = True  # pragma: no cover
+    except KeyError as e:
+        if "No bdist_wheel releases found" not in str(e):
```

Comment:
Am I understanding correctly that if a wheel isn't present we fall back to full install rather than parsing the `sdist`?

Reply:
correct.  I worked on parsing the sdist a bit... but it's a bit of a can of worms.  In a future PR, I do think we could speed up a few npe2 repos that do have manifests, but don't have wheels